File: java/src/net/razorvine/pyro/PyroURI.java
package net.razorvine.pyro;

import java.io.IOException;
import java.io.Serializable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * The Pyro URI object.
 * 
 * @author Irmen de Jong (irmen@razorvine.net)
 */
public class PyroURI implements Serializable {

	private static final long serialVersionUID = -7611447798373262153L;
	public String protocol = "PYRO";
	public String objectid;
	public String host;
	public int port;

	public PyroURI() {
	}

	public PyroURI(PyroURI other) {
		protocol = other.protocol;
		objectid = other.objectid;
		host = other.host;
		port = other.port;
	}

	public PyroURI(String uri) {
		Pattern p = Pattern.compile("(PYRO[A-Z]*):(\\S+?)(@(\\S+))?$");
		Matcher m = p.matcher(uri);
		if (m.find()) {
			protocol = m.group(1);
			objectid = m.group(2);
			String[] loc = m.group(4).split(":");
			host = loc[0];
			port = Integer.parseInt(loc[1]);
		} else {
			throw new PyroException("invalid URI string");
		}
	}

	public PyroURI(String objectid, String host, int port) {
		this.objectid = objectid;
		this.host = host;
		this.port = port;
	}

	public String toString() {
		return "<PyroURI " + protocol + ":" + objectid + "@" + host + ":" + port + ">";
	}


	@Override
	public int hashCode() {
		return toString().hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (!(obj instanceof PyroURI))
			return false;
		PyroURI other = (PyroURI) obj;
		return toString().equals(other.toString());
	}

	/**
	 * called by the Unpickler to restore state
	 */
	public void __setstate__(Object[] args) throws IOException {
		this.protocol = (String) args[0];
		this.objectid = (String) args[1];
		this.host = (String) args[3];
		this.port = (Integer) args[4];
	}
}

File: dotnet/Pyrolite/Pyro/PyroURI.cs
/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */

using System.Text.RegularExpressions;
using Razorvine.Pyrolite.Pickle;
using Razorvine.Pyrolite.Pickle.Objects;

namespace Razorvine.Pyrolite.Pyro
{

/// <summary>
/// The Pyro URI object.
/// </summary>
public class PyroURI {

	public string protocol {get;set;}
	public string objectid {get;set;}
	public string host {get;set;}
	public int port {get;set;}

	static PyroURI() {
		Unpickler.registerConstructor("Pyro4.core", "URI", new AnyClassConstructor(typeof(PyroURI)));
		Pickler.registerCustomPickler(typeof(PyroURI), new PyroUriPickler());
	}
	
	public PyroURI() {
		protocol="PYRO";
	}

	public PyroURI(PyroURI other) {
		protocol = other.protocol;
		objectid = other.objectid;
		host = other.host;
		port = other.port;
	}

	public PyroURI(string uri) {
		Regex r = new Regex("(PYRO[A-Z]*):(\\S+?)(@(\\S+))?$");
		Match m = r.Match(uri);
		if (m.Success) {
			protocol = m.Groups[1].Value;
			objectid = m.Groups[2].Value;
			string[] loc = m.Groups[4].Value.Split(':');
			host = loc[0];
			port = int.Parse(loc[1]);
		} else {
			throw new PyroException("invalid URI string");
		}
	}

	public PyroURI(string objectid, string host, int port) {
		this.objectid = objectid;
		this.host = host;
		this.port = port;
	}

	public override string ToString() {
		return "<PyroURI " + protocol + ":" + objectid + "@" + host + ":" + port + ">";
	}

	#region Equals and GetHashCode implementation
	public override bool Equals(object obj)
	{
		PyroURI other = obj as PyroURI;
		if (other == null)
			return false;
		return other.ToString()==this.ToString();
	}
	
	public override int GetHashCode()
	{
		return ToString().GetHashCode();
	}
	
	public static bool operator ==(PyroURI lhs, PyroURI rhs)
	{
		if (ReferenceEquals(lhs, rhs))
			return true;
		if (ReferenceEquals(lhs, null) || ReferenceEquals(rhs, null))
			return false;
		return lhs.Equals(rhs);
	}
	
	public static bool operator !=(PyroURI lhs, PyroURI rhs)
	{
		return !(lhs == rhs);
	}
	#endregion

	
	/**
	 * called by the Unpickler to restore state
	 */
	public void __setstate__(object[] args) {
		this.protocol = (string) args[0];
		this.objectid = (string) args[1];
		this.host = (string) args[3];
		this.port = (int) args[4];
	}
}

}
